### PR TITLE
feat(rust): add nostd + alloc compilation support to ockam

### PR DESCRIPTION
**Pull request by antoinevg**
### Proposed Changes

* fix(rust): no_std version of thread_rng state does not advance with repeated instantiations 
* fix(rust): match no_std semantics of compat::mutex lock implementation to std
* fix(rust): fix crate compilation for no_std + alloc feature configuration
  - ockam
  - ockam_entity
  - ockam_channel
  - ockam_vault*
  - signature*
  - ockam_node
* chore(rust): add ockam_core::compat::task
* chore(rust): use forked version of crates core2 and serde_bare

### CODE REVIEW

File: implementations/rust/ockam/ockam/src/protocols/parser.rs [hunk 1]
```diff
@@ -102,6 +106,9 @@ impl<W: Worker> ProtocolParserImpl<W> {
     {
         let p: Arc<Box<dyn ParserFragment<W> + Send + Sync>> = Arc::new(Box::new(parser));
 
+        #[cfg(not(feature = "std"))]
+        let mut map = self.map.write(); // TODO
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam/src/protocols/parser.rs:110]: What should be done here?

File: implementations/rust/ockam/ockam_node/src/error.rs [hunk 2]
```diff
@@ -1,5 +1,5 @@
+use crate::tokio::{self, sync::mpsc::error::SendError};
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_node/src/error.rs:1]: What is the difference here?

File: implementations/rust/ockam/ockam_node/src/relay/processor_relay.rs [hunk 3]
```diff
@@ -52,14 +52,15 @@ where
             Result::<()>::Ok(())
         };
 
+        #[allow(dead_code)] // TODO @antoinevg
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_node/src/relay/processor_relay.rs:55]: Can we fix it now?

**Comment by SanjoDeundiak** [comment]
Very nice!

File: implementations/rust/ockam/ockam_node/src/relay/processor_relay.rs (hunk 3 above)
Comment (antoinevg) [reply implementations/rust/ockam/ockam_node/src/relay/processor_relay.rs:55]: I feature-flagged it so It's fixed-ish enough for the embedded tech preview. Proper fix has to wait until either upstream stabilizes `select!()` for no_std or I implement our own.

File: implementations/rust/ockam/ockam/src/protocols/parser.rs [hunk 4]
```diff
@@ -120,7 +127,9 @@ impl<W: Worker> ProtocolParserImpl<W> {
         trace!("Parsing message for '{}' protocol", proto.as_str());
 
         // Get the protocol specific parser
-        let map = self.map.read().unwrap();
+        let map = self.map.read(); // TODO
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam/src/protocols/parser.rs:130]: Let's either remove TODO or better - add some details

File: implementations/rust/ockam/ockam_node/src/relay/processor_relay.rs (hunk 3 above)
Comment (SanjoDeundiak) [reply implementations/rust/ockam/ockam_node/src/relay/processor_relay.rs]: @antoinevg can you add those details to TODO?

File: implementations/rust/build.gradle [hunk 5]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (SanjoDeundiak) [inline implementations/rust/build.gradle:1]: Let's add this line back :)

Comment (antoinevg) [reply implementations/rust/build.gradle:1]: Ah, my editor is super-zealous about white-space removal! I've put it back where I found it.

**Review by jared-s — APPROVED** [review]
:partying_face:

**Pull request merged**